Add context for translating ShelleyLedgerConfig

bartfrenk opened the pull request:
Work-in-progress: Still needs to be branched off from head of master.

These changes branch of from the commit labeled ledger/1.1.0, which, at the time of writing, is the version that `master` of ouroboros-network depends on.

The aim is to close https://github.com/input-output-hk/ouroboros-consensus/issues/409, which is done by https://github.com/input-output-hk/ouroboros-network/pull/4091. That PR in ouroboros-network depends on changes in this PR.

Commit:
update changelog for 1.1.x release

also update the contributing document with our use of tags for patches.

nfrisby committed:
Add missing instances for Conway era:

* `GetLedgerView` for `ConwayEra` in `cardano-protocol-tpraos`
* `Eq`, `Generic`, `NoThunks` for `ConwayGenesis`

nfrisby committed:
Create a placeholder `Test.Cardano.Ledger.Conway.Serialisation.Generators`

lehins committed:
Make Arbitrary instances Era agnostic

lehins committed:
Switch larger functions from `INLINE` to `INLINEABLE` in order to reduce memory overhead during compilation

lehins committed:
Apply few hlint suggestions

lehins committed:
Fix setter for Allegra TxBody in tests

lehins committed:
Ensure GithubActions CI runs for release branches

Commit:
Merge pull request #2993 from input-output-hk/lehins/cherry-pick-some-of-recent-changes

Cherry pick some of recent changes

JaredCorduan commented on eras/shelley/impl/src/Cardano/Ledger/Shelley/Genesis.hs:
```diff
@@ -116,6 +119,16 @@ emptyGenesisStaking =
+  { pdsgSlotLength     :: !NominalDiffTime
+  , pdsgEpochLength    :: !EpochSize
+  , pdsgProtocolParams :: !(ShelleyPParams era)
+  } deriving (Eq, Show, Generic)
```
this is great to see, thanks for taking this on @bartfrenk !

JaredCorduan commented:
@bartfrenk should we close this for the time being?

bartfrenk commented:
I have put this on hold until @nfrisby was back from holidays, since this PR was part of resolving this issue https://github.com/input-output-hk/ouroboros-consensus/issues/409, which would involve changing the external API of consensus. It seems like something we would still want to do, and confirming that is next on my priority list. @JaredCorduan

nfrisby commented on cabal.project:
```diff
@@ -30,6 +30,9 @@ packages:
   libs/set-algebra
   libs/vector-map
 
+program-options
```
My understanding is that the typical policy among IOG repos is for each developer to add `-Werror` to their `cabal.project.local` file, if they prefer, instead of unilaterally putting it in the repository. I don't know for sure whether that's the case for `cardano-ledger`, but I would guess so.

lehins replied:
It is the opposite in ledger. We have `-Werror` enabled for all projects. If a developer wants to turn them back into warnings then he/she can do so in the `cabal.project.local`

nfrisby commented on eras/shelley-ma/impl/cardano-ledger-shelley-ma.cabal:
```diff
@@ -52,6 +52,9 @@ library
   other-modules:
     Cardano.Ledger.ShelleyMA.Era
     Cardano.Ledger.ShelleyMA.Rules.Utxo
+    Cardano.Ledger.ShelleyMA.Rules.Utxo
```
This looks like an accidental duplicate line?

nfrisby commented on eras/shelley/impl/src/Cardano/Ledger/Shelley/Era.hs:
```diff
@@ -1,4 +1,4 @@
-{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE DataKinds    #-}
```
I think `cardano-ledger` has different style rules than we do in Consensus. (If I recall correctly, they use `ormolu`.) So tweaks like this are probably out-of-bounds for this PR.

nfrisby commented on libs/cardano-ledger-core/src/Cardano/Ledger/Core.hs:
```diff
@@ -463,6 +464,8 @@ class
 -- > type instance PreviousEra (AllegraEra c) = ShelleyEra c
 type family PreviousEra era = (r :: Type) | r -> era
 
+type family ByronTranslationContext era :: Type
```
I see that you're instantiating this for eras other than `ShelleyEra`. That seems wrong to me. As far as I know, `ShelleyEra` is the only one that is designed to be capable of following `Byron`.

What do you think, @JaredCorduan ?

JaredCorduan replied:
That is correct @nfrisby , only `ShelleyEra` should be capable of following `Byron`, and is not a "core" feature for all eras.

nfrisby replied:
Ah, I realize now that @bartfrenk copied these instances from my sketch in the Issue input-output-hk/ouroboros-network#3976. I suppose me disagreeing with my past self shows the sketch needs more thought!

In the Consensus layer, we're reusing one data type for all the Shelley-based eras. For the Shelley era itself, that data type should contain whatever it needs to translate the final Byron ledger state into the first Shelley ledger state. But we want any other era to have `()` for that field, since no other era needs to translate the final Byron ledger state.

Another way to think about it: the `TranslationContext` for `ShelleyEra` should include the necessary bits for the Byron ledger state translation, whereas all the era Shelley-based eras wouldn't need those bits. However, `TranslationContext` is only for translating from Shelley-based era to Shelley-based era, and thus there is no `TranslationContext` for `ShelleyEra` (since it's the first Shelley-based era).

So: `type instance ByronTranslationContext (BabbageEra _) = ()` doesn't mean "`BabbageEra` can translate from Byron without needing any addition data." It instead means "`BabbageEra` doesn't need any such context data since it never translates from Byron".

Maybe a renaming would help clarify?

Alternatively, we could have this data type:

```
data ShelleyLedgerConfig era = ShelleyLedgerConfig {
      shelleyLedgerGlobals                     :: !SL.Globals
    , shelleyLedgerTranslationContext          :: !(Core.TranslationContext era)
    , shelleyLedgerTranslationContextFromByron :: !(Strict.Maybe (SL.ByronTranslationContext era))
    }

type instance ByronTranslationContext ShelleyEra = SL.ParedDownShelleyGenesis
type instance ByronTranslationContext AllegraEra = TypeError ...
type instance ByronTranslationContext AlonzoEra  = TypeError ...
type instance ByronTranslationContext BabbageEra = TypeError ...
type instance ByronTranslationContext ConwayEra  = TypeError ...
```

Thus the person constructing the `ShelleyLedgerConfig ` could `Just` only with the `ShelleyEra`, and must use `Nothing` with all the others, as intended. I _think_ this would also effectively prevent any of those non-`ShelleyEra` era's code from accessing this field: the type of the argument when checking a `Just x` pattern would raise a typeerror, wouldn't it?

Edit: Ah, but want the data type to definitely contain the `ByronTranslationContext` for the `ShelleyEra`--`Nothing` would be very problematic there. So I think we simply need a richer type for the family:

```
type family ByronTranslationContextIfAny era :: Maybe Type

data StaticMaybe :: Maybe Type -> Type where
  StaticJust    :: !a -> StaticMaybe (Just a)
  StaticNothing :: StaticMaybe Nothing
```

nfrisby replied:
Actually, I think the following is the much simpler and ideal option. Just redefine the already existing `TranslationContext (ShelleyEra c)` type family instance from its current `()` right-hand side to the following:

```
-- | There is no @TranslateEra (ShelleyEra c)@ instance, since 'TranslateEra' is for translating from one Shelley-based era to the next, and 'ShelleyEra' is the first such.
-- However, we do need extra data when translating from Byron to 'ShelleyEra' itself. Hence this instance.
type instance TranslationContext (ShelleyEra c) = TranslationContextByronToShelley c

data TranslationContextByronToShelley c = TranslationContextByronToShelley {
  byronToShelleyProtocolParams    :: ShelleyPParams (ShelleyEra c)
, byronToShelleyGenDelegs         :: GenDelegs c
, byronToShelleyMaxLovelaceSupply :: Word64
}
```

cc @amesgen

amesgen replied:
That indeed seems like the most minimal change that would do the job. :+1:

Obviously, this would special-case Byron as a predecessor of the first Shelley era, as we store the `TranslationContext` directly in the `ShelleyLedgerConfig` in Consensus. I am not sure how keen Ledger is on organizing the code in such a way that the Cardano chain (which includes Byron) is not given special treatment compared to a Shelley-only chain (where `TranslationContext (ShelleyEra c) ~ ()` would be natural) or even a chain where Shelley is preceded by a protocol other than Byron (on which `TranslationContext (ShelleyEra c)` might depend).

At least e.g. the docs of `ShelleyGenesis` are alluding to the possibility of a Shelley-only chain:
https://github.com/input-output-hk/cardano-ledger/blob/836452c1ff7b38b16ca5d2d3b0c58383ef59f512/eras/shelley/impl/src/Cardano/Ledger/Shelley/Genesis.hs#L121-L125

JaredCorduan replied:
@amesgen I believe that there are tests from the QA team, and benchmarks from the benchmarking, that make use of a Shelley-only chain. We should check with them.

> I am not sure how keen Ledger is on organizing the code in such a way that the Cardano chain (which includes Byron) is not given special treatment ...

I can't say that I understand the implications of this, sorry.

nfrisby reviewed (commented):
Just passing by, had these thoughts to share.

bartfrenk committed:
Add non-trivial TranslationContext for Shelley, remove TranslateEra * ShelleyGenesis instances

To support the ouroboros-network PR here:
https://github.com/input-output-hk/ouroboros-network/pull/4091
